### PR TITLE
Add basic Datapack API (#5653)

### PR DESCRIPTION
**Pull request by ConnorLinfoot**
Adds an API to be able to interface with the resource packs on the server (also known as data packs) including being able to get a list of packs and if they are enabled, and be able to enable or disable them.

### CODE REVIEW

File: Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
++    private final ResourcePackRepository repository;
++
++    public PaperResourcePackManager(ResourcePackRepository repository) {
++        this.repository = repository;
++    }
++
++    @Override
++    public Collection<ResourcePack> getPacks() {
++        Collection<ResourcePackLoader> enabledPacks = repository.e();
++        return repository.c().stream().map(loader -> new PaperResourcePack(loader, enabledPacks.contains(loader))).collect(Collectors.toList());
```
Comment (kennytv) [inline Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch]: Obfhelpers

File: Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
++    }
++
++    @Override
++    public Collection<ResourcePack> getEnabledPacks() {
++        return repository.e().stream().map(loader -> new PaperResourcePack(loader, true)).collect(Collectors.toList());
```
Comment (kennytv) [inline Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch]: Obfhelper

File: Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
++import java.util.stream.Collectors;
++
++public class PaperResourcePack implements ResourcePack {
++    private final String name;
++    private final ResourcePackCompatibility compatibility;
++    private final boolean enabled;
++
++    PaperResourcePack(ResourcePackLoader loader, boolean enabled) {
++        this.name = loader.e();
++        this.compatibility = ResourcePackCompatibility.valueOf(loader.c().name());
```
Comment (kennytv) [inline Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch]: Obfhelpers

File: Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
++    }
++
++    @Override
++    public void setEnabled(boolean enabled) {
++        if (enabled == this.enabled) {
++            return;
++        }
++
++        MinecraftServer server = MinecraftServer.getServer();
++        List<String> enabledKeys = server.getResourcePackRepository().e().stream().map(ResourcePackLoader::e).collect(Collectors.toList());
```
Comment (kennytv) [inline Spigot-Server-Patches/0731-Add-basic-ResourcePack-DataPack-API.patch]: Obfhelpers

**Review by kennytv — DISMISSED** [review]
Needs some obfuscation helpers, see https://github.com/PaperMC/Paper/blob/master/CONTRIBUTING.md#obfuscation-helpers

File: Spigot-API-Patches/0300-Add-basic-ResourcePack-DataPack-API.patch [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
++}
+diff --git a/src/main/java/io/papermc/paper/resource/ResourcePackCompatibility.java b/src/main/java/io/papermc/paper/resource/ResourcePackCompatibility.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..f0f6e28754df2c552c0f139ea9d7f50e65ad6a6a
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/resource/ResourcePackCompatibility.java
+@@ -0,0 +1,7 @@
++package io.papermc.paper.resource;
++
++public enum ResourcePackCompatibility {
```
Comment (Machine-Maker) [inline Spigot-API-Patches/0300-Add-basic-ResourcePack-DataPack-API.patch]: I'd probably throw this enum inside the ResourcePack interface, kinda small and specific for its own file.

Can just call it `Compatibility` then too.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Comment by ConnorLinfoot** [comment]
That should be all the Obfhelpers added and the compatibility enum change :)

File: Spigot-API-Patches/0300-Add-basic-ResourcePack-DataPack-API.patch [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+@@ -1937,6 +1939,14 @@ public final class Bukkit {
+     public static com.destroystokyo.paper.entity.ai.MobGoals getMobGoals() {
+         return server.getMobGoals();
+     }
++
++    /**
++     * @return the resource pack manager
++     */
++    @NotNull
++    public static ResourcePackManager getResourcePackManager() {
```
Comment (JRoy) [inline Spigot-API-Patches/0300-Add-basic-ResourcePack-DataPack-API.patch]: please fully qualify this and remove the import

File: Spigot-API-Patches/0300-Add-basic-ResourcePack-DataPack-API.patch [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+@@ -1698,5 +1700,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
+      */
+     @NotNull
+     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
++
++    /**
++     * @return the resource pack manager
++     */
++    @NotNull
++    ResourcePackManager getResourcePackManager();
```
Comment (JRoy) [inline Spigot-API-Patches/0300-Add-basic-ResourcePack-DataPack-API.patch]: fully qualify

**Comment by MiniDigger** [comment]
looks good to me, only concern I have is if we really wanna name this resource pack, as ppl kinda expect this to be called data pack? I know that the mappings use the resource name, but mojang doesn't, they just call it pack in their code.
there is one other PR in this area, https://github.com/PaperMC/Paper/pull/5021, that also uses the datapack name.
like, just that fact that you felt like you needed to add a "aka data packs" to the title shows that the resource pack name is misleading. 

what do others think?

**Comment by Machine-Maker** [comment]
Yeah, I would call them datapacks, (idk about capitalization). Wiki uses datapacks, and resource packs already mean something else (the packs clients use to change textures, sounds, etc).

**Comment by ConnorLinfoot** [comment]
Yeah I agree on the naming being better as a data pack api rather than developers being confused and maybe even missing the existence due to the bad naming

**Review by JRoy — DISMISSED** [review]

**Review by Proximyst — APPROVED** [review]

**Comment by ConnorLinfoot** [comment]
Resolved the above issue and went ahead and renamed it to be datapack

File: Spigot-API-Patches/0300-Add-basic-Datapack-API.patch [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
++}
+diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
+index 5f7208196684d9c8373df28b7cfb5f9e21baa41e..36405c6e588c9e5935d22a6e54ed6410da0c10ba 100644
+--- a/src/main/java/org/bukkit/Bukkit.java
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -13,6 +13,7 @@ import java.util.Set;
+ import java.util.UUID;
+ import java.util.function.Consumer;
+ import java.util.logging.Logger;
++
```
Comment (JRoy) [inline Spigot-API-Patches/0300-Add-basic-Datapack-API.patch]: :P
```suggestion
```

File: Spigot-API-Patches/0300-Add-basic-Datapack-API.patch [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+     @NotNull
+diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
+index f3e27d2d02a9407bb1b091b8c1125ad5abf99e55..6fb924b838dd379024560f124aaf7bb6dcbfb937 100644
+--- a/src/main/java/org/bukkit/Server.java
++++ b/src/main/java/org/bukkit/Server.java
+@@ -13,6 +13,7 @@ import java.util.Set;
+ import java.util.UUID;
+ import java.util.function.Consumer;
+ import java.util.logging.Logger;
++
```
Comment (JRoy) [inline Spigot-API-Patches/0300-Add-basic-Datapack-API.patch]:
```suggestion
```

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0300-Add-basic-Datapack-API.patch [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+index f3e27d2d02a9407bb1b091b8c1125ad5abf99e55..6fb924b838dd379024560f124aaf7bb6dcbfb937 100644
+--- a/src/main/java/org/bukkit/Server.java
++++ b/src/main/java/org/bukkit/Server.java
+@@ -13,6 +13,7 @@ import java.util.Set;
+ import java.util.UUID;
+ import java.util.function.Consumer;
+ import java.util.logging.Logger;
+ import org.bukkit.Warning.WarningState;
+ import org.bukkit.advancement.Advancement;
+ import org.bukkit.block.data.BlockData;
```
Comment (JRoy) [inline Spigot-API-Patches/0300-Add-basic-Datapack-API.patch]:
```suggestion
```

File: Spigot-API-Patches/0300-Add-basic-Datapack-API.patch [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+index 5f7208196684d9c8373df28b7cfb5f9e21baa41e..36405c6e588c9e5935d22a6e54ed6410da0c10ba 100644
+--- a/src/main/java/org/bukkit/Bukkit.java
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -13,6 +13,7 @@ import java.util.Set;
+ import java.util.UUID;
+ import java.util.function.Consumer;
+ import java.util.logging.Logger;
+ import org.bukkit.Warning.WarningState;
+ import org.bukkit.advancement.Advancement;
+ import org.bukkit.block.data.BlockData;
```
Comment (JRoy) [inline Spigot-API-Patches/0300-Add-basic-Datapack-API.patch]:
```suggestion
```

**Review by JRoy — APPROVED** [review]

**Comment by JRoy** [comment]
My bad, I didn't expect you to actually commit the code suggestions on github lmao, the patch may fail to apply now?

**Comment by ConnorLinfoot** [comment]
> My bad, I didn't expect you to actually commit the code suggestions on github lmao, the patch may fail to apply now?

Seems to be fine :)

**Review by MiniDigger — APPROVED** [review]

**Comment by JRoy** [comment]
Thanks for the contribution :)

**Comment by kennytv** [comment]
In case you or someone else is already depending on this; some jd and package "resource" references will be renamed with https://github.com/PaperMC/Paper/pull/5662

**Pull request merged**